torchao/prototype/float8nocompile/test/test.py

```diff
@@ -0,0 +1,74 @@
+import pytest
+import torch
+import torch.nn as nn
+
+from torchao.float8.float8_linear_utils import convert_to_float8_training
+from torchao.prototype.float8nocompile.float8nocompile_linear_utils import (
+    convert_to_float8_nocompile_training,
+)
+from torchao.utils import TORCH_VERSION_AT_LEAST_2_5
+
+if not TORCH_VERSION_AT_LEAST_2_5:
+    raise AssertionError("torchao.float8 requires PyTorch version 2.5 or greater")
+
+
+class TestModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.layers = nn.Sequential(
+            nn.Linear(32, 64, bias=False),
+            nn.Linear(64, 32, bias=False),
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.layers(x)
+
+
+@pytest.fixture
+def model1():
+    torch.manual_seed(0)
+    return TestModel()
+
+
+@pytest.fixture
+def model2():
+    torch.manual_seed(0)
+    return TestModel()
+
+
+def test_model_weights_and_gradients(model1, model2):
+    assert torch.cuda.is_available()
+    device = torch.device("cuda")
+
+    model1.to(torch.bfloat16).to(device)
+    model2.to(torch.bfloat16).to(device)
+
+    # swap nn.Linear layers with Float8LinearNoCompile layers in model1 only
+    convert_to_float8_nocompile_training(model1)
+
+    input_data = torch.randn(16, 32, dtype=torch.bfloat16).to(device)
+    optim1 = torch.optim.SGD(model1.parameters(), lr=0.01)
+    optim2 = torch.optim.SGD(model2.parameters(), lr=0.01)
+    loss_fn = nn.MSELoss()
+
+    for i in range(1):
+        print(f"step {i}")
+        optim1.zero_grad()
+        optim2.zero_grad()
+
+        output1 = model1(input_data)
+        output2 = model2(input_data)
+
+        loss1 = loss_fn(output1, torch.zeros_like(output1))
+        loss2 = loss_fn(output2, torch.zeros_like(output2))
+
+        loss1.backward()
+        loss2.backward()
+
+        optim1.step()
+        optim2.step()
+
+        # compare the weights and gradients of both models
+        for param1, param2 in zip(model1.parameters(), model2.parameters()):
+            assert torch.allclose(param1.data, param2.data, atol=1e-3, rtol=1e-3)
```

Comment:
I think it should match with atol=0 and rtol=0, does that work?

You can also construct a simpler test to test the same thing with atol=0 and rtol=0:
1. start with sample input
2. do one forward and one backward (no optimizer update)
3. compare output, weight gradient, input gradient

If there is a divergence, the setup above will be easier to debug compared to a full training loop

Reply:
Yeah I used 1e-3 because when I tested with 1e-5 there was divergence. I will look into it further.

Reply:
As I test I used the official `convert_to_float8_training` implementation rather than my own, and there is still divergence, specifically on the weight gradients

Reply:
Actually I'm wondering if a slight amount of divergence in model outputs (and thus gradients) is expected when converting from bf16 to fp8, since we are using a lower precision data type there will be some amount of rounding error during the conversion, right?

Reply:
Ah, was your comment comment was specifically talking about requiring an exact match for `param.data`, not `param.gradient` as well? If so then this makes sense and I've updated the code accordingly.

I also removed the optimizers since they aren't needed to verify the weights and gradients.

Reply:
if your first model is converted to float8 with the production path with default settings and your second model is converted to float8 with your path, then you should hopefully be able to match bitwise between them.  There isn't a strong need to compare to a bfloat16 baseline here.

Reply:
I see, I misunderstood what we wanted to compare here. That makes sense, I've updated the code accordingly to compare production float8 linear conversion with no-compile version and verify the param data and grads are an exact match.